### ISSUE TITLE
Unstable inference results with c++

### ISSUE THREAD

**Issue by zhuxiaoxuhit** [microsoft/onnxruntime#5600]
**Describe the bug**
I converted tensorflow model (tacotron) to onnx with c++ on cpu. Everything is going well. But a phenomenon puzzled me : I created session once, then made inferences with same input repeatedly, but got different results. 
Is there something random somewhere in onnxruntime?
ps: I used dropout in inference, but set a fixed seed value in advance.

**System information**
- OS Platform and Distribution:CentOS Linux release 7.5.1804
- ONNX Runtime installed from (source or binary): source
- ONNX Runtime version: master
- Python version: 3.6

**Expected behavior**
Get same results everytime inference.

**Screenshots**
![image](https://user-images.githubusercontent.com/32813150/97276987-de4f2800-1872-11eb-98c9-dbb38159f99e.png)

**Comment by yuslepukhin** [comment]
Would you be able to share the model and the commit you used to build the product off?

**Comment by freedenS** [comment]
@yuslepukhin
Met the same problem.
I have tested ORT-1.3.0, 1.4.0 and 1.5.2 with python & c++.
Here is my model and image. https://drive.google.com/file/d/1XpLTfNpqEdBVi1glnHFtMDvdDT38C7Tn/view?usp=sharing
Here is my python code.
```
import onnxruntime
import cv2
import torch
sess = onnxruntime.InferenceSession('./test.onnx')
inputs = cv2.imread('1.jpg')
inputs = cv2.resize(inputs, (800,1067))
inputs = torch.as_tensor(inputs.astype('float32').transpose(2,0,1))
inputs = inputs.unsqueeze(0)
for i in range(100):
    res = sess.run(None, {sess.get_inputs()[0].name:inputs.numpy()})
    print (res[0].shape[0])
```

**Comment by zhuxiaoxuhit** [comment]
@yuslepukhin Is this bug fixed? I want to share some information. I compared params ’Training‘ in tensorflow dropout,
| params/results  | Run 1 times with 1 session | Run many times with 1 session |
| ----------------|:---------------:| --------:|
| Training=True  | diff  | diff  |
| Training=False | same | diff  |

**Comment by yuslepukhin** [comment]
@zhuxiaoxuhit Thank you for this datapoint.

**Comment by yuslepukhin** [comment]
@zhuxiaoxuhit The issue title uses the term `stable`, however, later in the comment you mention the same results. We can't guarantee the latter with the floating point arithmetic. Looking at the chart you published it looks very similar with some time shift. What is your tolerance? We often can not produce more precision than 3 digits after decimal point.
Some output examples would be helpful.

**Comment by zhuxiaoxuhit** [comment]
@yuslepukhin "STABLE" means that:

**Comment by yuslepukhin** [comment]
Here is the documentation for [Dropout](https://github.com/onnx/onnx/blob/master/docs/Operators.md#Dropout).
If you run it in training mode, there will be a randomized dropout.  In inference mode  where output is a copy of the input nothing is dropped out and our optimizer eliminates Dropout node from the execution graph. So if you are running an inference, the issue is probably not in the dropout.

**Comment by yuslepukhin** [comment]
I have not answered your question fully. Seed is an attribute of Dropout and as such it is a part of the serialized model, which is protobuf. At runtime, attributes are const. If you want to modify it you would have to edit the model and protobuf has many APIs to accomplish that, including python and C++.

**Comment by yuslepukhin** [comment]
I just built a Debug build on Windows  off c9f44276da0919c4c7e7409210142beb61c8ab11 master. And ran the above script and the model. I have got 100 identical results.  I have not tried Linux yet as I was short on time.

6
[[ 87.6838  807.0981  229.68369 942.5767 ]
 [442.69327 282.0529  538.2384  343.02744]
 [381.2425  297.12918 426.899   317.2443 ]
 [268.38806 454.03195 299.77087 547.7364 ]
 [381.59628 275.90866 424.5983  343.25797]
 [444.21994 284.96368 490.20328 344.88654]]
6
[[ 87.6838  807.0981  229.68369 942.5767 ]
 [442.69327 282.0529  538.2384  343.02744]
 [381.2425  297.12918 426.899   317.2443 ]
 [268.38806 454.03195 299.77087 547.7364 ]
 [381.59628 275.90866 424.5983  343.25797]
 [444.21994 284.96368 490.20328 344.88654]]
6
[[ 87.6838  807.0981  229.68369 942.5767 ]
 [442.69327 282.0529  538.2384  343.02744]
 [381.2425  297.12918 426.899   317.2443 ]
 [268.38806 454.03195 299.77087 547.7364 ]
 [381.59628 275.90866 424.5983  343.25797]
 [444.21994 284.96368 490.20328 344.88654]]
...

**Comment by freedenS** [comment]
@yuslepukhin have you tried this on cpu? It is stable on cpu. The problem will be happen only under gpu mode.

**Comment by yuslepukhin** [comment]
@freedenS This specific ticket talks about differences on CPU. Although the example is yours, I would appreciate the feedback from @zhuxiaoxuhit.

**Comment by yuslepukhin** [comment]
Tested optimized build on CPU, works the same.
Tested Debug build on GPU, shows identical results.
6
[[ 87.68379 807.0981  229.68367 942.5767 ]
 [442.69327 282.05286 538.2384  343.0274 ]
 [381.2425  297.12918 426.899   317.2443 ]
 [268.38806 454.03192 299.77087 547.7363 ]
 [381.59628 275.90866 424.5983  343.25797]
 [444.21994 284.96368 490.20328 344.88654]]

I have a suggestion. Perhaps, the issue was fixed? Happens all the time. Try that commit for yourself.

**Comment by freedenS** [comment]
@yuslepukhin Thanks for your help! I'll take a try.
